lib/SILOptimizer/Mandatory/Differentiation.cpp

```diff
@@ -4153,87 +3965,1112 @@ inline llvm::raw_ostream &operator<<(llvm::raw_ostream &os,
+    // TODO: Handle indirect results.
+    auto tanParam =
+        materializeTangent(getTangentValue(ri->getOperand()), loc);
+    diffBuilder.createReturn(ri->getLoc(), tanParam);
+  }
+
+  void emitTangentForApplyInst(ApplyInst *ai, SILAutoDiffIndices indices) {
+    auto *bb = ai->getParent();
+    auto loc = ai->getLoc();
+    auto diffBuilder = getDifferentialBuilder();
+
+    // Get the differential.
+    auto *field = differentialInfo.lookUpLinearMapDecl(ai);
+    assert(field);
+    SILValue differential = getDifferentialStructElement(bb, field);
+
+    SmallVector<SILValue, 8> diffArgs;
+    for (auto origArg : ai->getArguments()) {
+      // Get the tangent value of the original parameter.
+      if (!activityInfo.isActive(origArg, getIndices()))
+        continue;
+      SILValue tanParam;
+      if (origArg->getType().isObject()) {
+        // If original result is a `tuple_extract`, materialize tangent value of
+        // `ai` and extract the corresponding element tangent value.
+        if (auto *tupleExtract =
+                dyn_cast<TupleExtractInst>(origArg)) {
+          auto tangentTuple = materializeTangent(getTangentValue(ai), loc);
+          tanParam = diffBuilder.emitTupleExtract(
+              loc, tangentTuple, tupleExtract->getFieldNo());
+        }
+        // Otherwise, materialize tangent value of `ai`.
+        else {
+          tanParam = materializeTangent(getTangentValue(origArg), loc);
+        }
+      } else {
+        tanParam = getTangentBuffer(ai->getParent(), origArg);
+        if (errorOccurred)
+          return;
+      }
+      diffArgs.push_back(tanParam);
+    }
+
+    // Call the differential.
+    auto *differentialCall = diffBuilder.createApply(
+        loc, differential, SubstitutionMap(), diffArgs,
+        /*isNonThrowing*/ false);
+    diffBuilder.emitDestroyValueOperation(loc, differential);
+    assert(differentialCall->getNumResults() == 1 &&
+           "Expected differential to return one result");
+
+    // TODO: Generalize for indirect results, multiple results, etc
+    auto origResult = ai->getResult(indices.source);
+
+    // Extract all direct results from the differential.
+    SmallVector<SILValue, 8> differentialDirResults;
+    extractAllElements(differentialCall, diffBuilder, differentialDirResults);
+    // Get all differential results in type-defined order.
+    SmallVector<SILValue, 8> differentialAllResults;
+    collectAllActualResultsInTypeOrder(
+        differentialCall, differentialDirResults,
+        differentialCall->getIndirectSILResults(), differentialAllResults);
+    auto differentialResult = differentialAllResults[indices.source];
+
+    // Add tangent for original result.
+    assert(indices.source == 0 && "Expected result index to be first.");
+    setTangentValue(bb, origResult,
+                    makeConcreteTangentValue(differentialResult));
+  }
+
+  void startDifferentialGeneration() {
```

Comment:
This is not starting differential generation, it's merely preparing for it, right?
1. Rename to `prepareForDifferentialGeneration`.
2. Add detailed doc comment to describe what this function does.
3. Put it next to `run()`.